File: LeetCode/Unknown/프로그래머스/4/17685. ［3차］ 자동완성/［3차］ 자동완성.java
import java.util.*;

class Solution {

    static class Word {
        char c;
        Map<Character, Word> nexts = new HashMap<>();
        int count = 0;

        Word(char c) {
            this.c = c;
        }
    }

    private void insert(Word cur, String word) {
        for (char ch : word.toCharArray()) {
            cur.nexts.putIfAbsent(ch, new Word(ch));
            cur = cur.nexts.get(ch);
            cur.count++;
        }
    }

    private int countType(Word cur, String word) {
        int cnt = 0;
        for (char ch : word.toCharArray()) {
            cnt++;
            cur = cur.nexts.get(ch);
            if (cur.count == 1) break;
        }
        return cnt;
    }

    public int solution(String[] words) {
        Word root = new Word(' ');
        for (String word : words) {
            insert(root, word);
        }

        int answer = 0;
        for (String word : words) {
            answer += countType(root, word);
        }

        return answer;
    }
}

File: 프로그래머스/4/17685. ［3차］ 자동완성/［3차］ 자동완성.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

struct Word {
    char c;
    map<char, Word*> nexts;
    int count = 0;  // 이 노드를 지나간 단어 개수
};

void insert(Word* cur, string word) {
    for (char ch : word) {
        if (cur->nexts.find(ch) == cur->nexts.end()) 
            cur->nexts[ch] = new Word{ch};
        
        cur = cur->nexts[ch];
        cur->count++;
    }
}

int countType(Word* cur, string word) {
    int cnt = 0;
    for (char ch : word) {
        cnt++;
        cur = cur->nexts[ch];
        
        if (cur->count == 1) break; // 유일해지는 순간
    }
    return cnt;
}

int solution(vector<string> words) {
    Word* root = new Word{' '};
    for (const string& word : words)
        insert(root, word);

    int answer = 0;
    for (string word : words)
        answer += countType(root, word);

    return answer;
}